lib/Importer/ONNXModelLoader.cpp

```diff
@@ -2332,9 +2332,33 @@ Error ONNXModelLoader::loadMatMul(const ONNX_NAMESPACE::NodeProto &op,
   NodeValue RHS;
   ASSIGN_VALUE_OR_RETURN_ERR(RHS, getNodeValueByName(op.input(1)));
 
-  /// For dimension size equal to 3 use batchedMatMul
-  if (LHS.dims().size() == 3) {
+  /// For dimension greater than 2 use batchedMatMul
+  if (LHS.dims().size() > 2) {
+    dim_t numLHSDims = LHS.dims().size();
+    dim_t numRHSDims = RHS.dims().size();
+    std::vector<dim_t> finalShape;
+    for (auto d : LHS.dims()) {
+      finalShape.push_back(d);
+    }
+    if (LHS.dims().size() > 3) {
+      RETURN_ERR_IF_NOT(numLHSDims == numRHSDims,
+                        "number of LHS and RHS dims should be equal");
+      finalShape[numLHSDims - 1] = RHS.dims()[numRHSDims - 1];
+      std::vector<dim_t> newLHSShape = {0, LHS.dims()[numLHSDims - 2],
+                                        LHS.dims()[numLHSDims - 1]};
+      newLHSShape[0] =
+          LHS.getType()->size() / (newLHSShape[1] * newLHSShape[2]);
+      LHS = G_->createReshape(opName, LHS, newLHSShape);
+      std::vector<dim_t> newRHSShape = {0, RHS.dims()[numRHSDims - 2],
+                                        RHS.dims()[numRHSDims - 1]};
+      newRHSShape[0] =
+          RHS.getType()->size() / (newRHSShape[1] * newRHSShape[2]);
+      RHS = G_->createReshape(opName, RHS, newRHSShape);
+    }
     Node *node = G_->createBatchMatMul(opName, LHS, RHS);
+    if (numLHSDims > 3) {
+      node = G_->createReshape(opName, node, finalShape);
+    }
```

Comment:
Would it make sense to just move all of this logic into `Function::createBatchMatMul()`? We already do some checking of LHS/RHS dimensions in there. Then we could also make an OperatorTest that calls `createBatchMatMul()` with 4D inputs.

Reply:
Yeah, I have moved the logic into Function::createBatchMatMul(). However, I have retained the final reshape (after batchmatmul) here itself as createBatchMatMul() returns batchmatmul node (we can't return reshape node there).